webapp/components/FilterPosts/FilterPosts.vue

```diff
@@ -0,0 +1,61 @@
+<template>
+  <dropdown ref="menu" :placement="placement" :offset="offset">
+    <a slot="default" slot-scope="{ toggleMenu }" href="#" @click.prevent="toggleMenu()">
+      <ds-icon style="margin: 12px 0px 0px 10px;" name="filter" size="large" />
+      <ds-icon style="margin: 7px 0px 0px 2px" size="xx-small" name="angle-down" />
+    </a>
+    <template slot="popover">
+      <filter-posts-menu-items :chunk="chunk" @filterPosts="filterPosts" />
+    </template>
+  </dropdown>
+</template>
+<script>
+import _ from 'lodash'
+import Dropdown from '~/components/Dropdown'
+import { filterPosts } from '~/graphql/PostQuery.js'
+import { mapMutations } from 'vuex'
+import FilterPostsMenuItems from '~/components/FilterPosts/FilterPostsMenuItems'
+
+export default {
+  components: {
+    Dropdown,
+    FilterPostsMenuItems,
+  },
+  props: {
+    placement: { type: String },
+    offset: { type: [String, Number] },
+    categories: { type: Array, default: () => [] },
+  },
+  data() {
+    return {
+      pageSize: 12,
+    }
+  },
+  computed: {
+    chunk() {
+      return _.chunk(this.categories, 2)
+    },
+  },
+  methods: {
+    ...mapMutations({
+      setPosts: 'posts/SET_POSTS',
+    }),
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-09T21:47:21Z" title="Tuesday, July 9th 2019, 11:47:21 pm +02:00">Jul 9, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,282 @@
+<template>
+  <dropdown ref="menu" :placement="placement" :offset="offset">
+    <a slot="default" slot-scope="{ toggleMenu }" href="#" @click.prevent="toggleMenu()">
+      <ds-icon style="margin: 5px 0px 0px 10px;" name="filter" size="large" />
+      <ds-icon style="margin-left: 2px" size="xx-small" name="angle-down" />
+    </a>
+    <template slot="popover">
+      <ds-container>
+        <ds-space />
+        <ds-flex :groupedCategories="chunk">
+          <ds-flex-item
+            :width="{ base: '100%', sm: '100%', md: '100%', lg: '20%' }"
+            class="categories-list"
+          >
+            <ds-space margin-bottom="x-small" />
+            <ds-flex id="filter-posts-header">
+              <ds-heading tag="h4">{{ $t('filter-posts.header') }}</ds-heading>
+              <ds-flex-item width="10%" />
+              <ds-flex-item width="100%">
+                <ds-button
+                  icon="check"
+                  @click.stop.prevent="toggleCategory()"
+                  :primary="allCategories"
+                />
+                <ds-flex-item>
+                  <label>{{ $t('filter-posts.all') }}</label>
+                </ds-flex-item>
+                <ds-space />
+              </ds-flex-item>
+            </ds-flex>
+          </ds-flex-item>
+          <ds-flex-item :width="{ base: '50%', sm: '0%', md: '50%', lg: '10%' }">
+            <ds-flex v-for="category in chunk[0]" :key="category.id" class="categories-list">
+              <ds-flex>
+                <ds-flex-item width="100%">
+                  <ds-button
+                    :icon="category.icon"
+                    :primary="isActive(category.id)"
+                    @click.stop.prevent="toggleCategory(category.id)"
+                  />
+                  <ds-space margin-bottom="small" />
+                </ds-flex-item>
+                <ds-flex>
+                  <ds-flex-item>
+                    <label>{{ category.name }}</label>
+                  </ds-flex-item>
+                  <ds-space margin-bottom="xx-large" />
+                </ds-flex>
+              </ds-flex>
+            </ds-flex>
+          </ds-flex-item>
+          <ds-flex-item :width="{ base: '50%', sm: '0%', md: '50%', lg: '10%' }">
+            <ds-flex v-for="category in chunk[1]" :key="category.id" class="categories-list">
+              <ds-flex>
+                <ds-flex-item width="100%">
+                  <ds-button
+                    :icon="category.icon"
+                    :primary="isActive(category.id)"
+                    @click.stop.prevent="toggleCategory(category.id)"
+                  />
+                  <ds-space margin-bottom="small" />
+                </ds-flex-item>
+                <ds-flex class="categories-list">
+                  <ds-flex-item>
+                    <label>{{ category.name }}</label>
+                  </ds-flex-item>
+                  <ds-space margin-bottom="xx-large" />
+                </ds-flex>
+              </ds-flex>
+            </ds-flex>
+          </ds-flex-item>
+          <ds-flex-item :width="{ base: '50%', sm: '0%', md: '50%', lg: '10%' }">
+            <ds-flex v-for="category in chunk[2]" :key="category.id" class="categories-list">
+              <ds-flex>
+                <ds-flex-item width="100%">
+                  <ds-button
+                    :icon="category.icon"
+                    :primary="isActive(category.id)"
+                    @click.stop.prevent="toggleCategory(category.id)"
+                  />
+                  <ds-space margin-bottom="small" />
+                </ds-flex-item>
+                <ds-flex>
+                  <ds-flex-item>
+                    <label>{{ category.name }}</label>
+                  </ds-flex-item>
+                  <ds-space margin-bottom="xx-large" />
+                </ds-flex>
+              </ds-flex>
+            </ds-flex>
+          </ds-flex-item>
+          <ds-flex-item :width="{ base: '50%', sm: '0%', md: '50%', lg: '10%' }">
+            <ds-flex v-for="category in chunk[3]" :key="category.id" class="categories-list">
+              <ds-flex>
+                <ds-flex-item width="100%">
+                  <ds-button
+                    :icon="category.icon"
+                    :primary="isActive(category.id)"
+                    @click.stop.prevent="toggleCategory(category.id)"
+                  />
+                  <ds-space margin-bottom="small" />
+                </ds-flex-item>
+                <ds-flex>
+                  <ds-flex-item>
+                    <label>{{ category.name }}</label>
+                  </ds-flex-item>
+                  <ds-space margin-bottom="xx-large" />
+                </ds-flex>
+              </ds-flex>
+            </ds-flex>
+          </ds-flex-item>
+          <ds-flex-item :width="{ base: '50%', sm: '0%', md: '50%', lg: '10%' }">
+            <ds-flex v-for="category in chunk[4]" :key="category.id" class="categories-list">
+              <ds-flex>
+                <ds-flex-item width="100%">
+                  <ds-button
+                    :icon="category.icon"
+                    :primary="isActive(category.id)"
+                    @click.stop.prevent="toggleCategory(category.id)"
+                  />
+                  <ds-space margin-bottom="small" />
+                </ds-flex-item>
+                <ds-flex>
+                  <ds-flex-item>
+                    <label>{{ category.name }}</label>
+                  </ds-flex-item>
+                  <ds-space margin-bottom="xx-large" />
+                </ds-flex>
+              </ds-flex>
+            </ds-flex>
+          </ds-flex-item>
+          <ds-flex-item :width="{ base: '50%', sm: '0%', md: '50%', lg: '10%' }">
+            <ds-flex v-for="category in chunk[5]" :key="category.id" class="categories-list">
+              <ds-flex>
+                <ds-flex-item width="100%">
+                  <ds-button
+                    :icon="category.icon"
+                    :primary="isActive(category.id)"
+                    @click.stop.prevent="toggleCategory(category.id)"
+                  />
+                  <ds-space margin-bottom="small" />
+                </ds-flex-item>
+                <ds-flex>
+                  <ds-flex-item>
+                    <label>{{ category.name }}</label>
+                  </ds-flex-item>
+                  <ds-space margin-bottom="xx-large" />
+                </ds-flex>
+              </ds-flex>
+            </ds-flex>
+          </ds-flex-item>
+          <ds-flex-item :width="{ base: '50%', sm: '0%', md: '50%', lg: '10%' }">
+            <ds-flex v-for="category in chunk[6]" :key="category.id" class="categories-list">
+              <ds-flex>
+                <ds-flex-item width="100%">
+                  <ds-button
+                    :icon="category.icon"
+                    :primary="isActive(category.id)"
+                    @click.stop.prevent="toggleCategory(category.id)"
+                  />
+                  <ds-space margin-bottom="small" />
+                </ds-flex-item>
+                <ds-flex>
+                  <ds-flex-item>
+                    <label>{{ category.name }}</label>
+                  </ds-flex-item>
+                  <ds-space margin-bottom="xx-large" />
+                </ds-flex>
+              </ds-flex>
+            </ds-flex>
+          </ds-flex-item>
+          <ds-flex-item :width="{ base: '50%', sm: '0%', md: '50%', lg: '10%' }">
+            <ds-flex v-for="category in chunk[7]" :key="category.id" class="categories-list">
+              <ds-flex>
+                <ds-flex-item width="100%">
+                  <ds-button
+                    :icon="category.icon"
+                    :primary="isActive(category.id)"
+                    @click.stop.prevent="toggleCategory(category.id)"
+                  />
+                  <ds-space margin-bottom="small" />
+                </ds-flex-item>
+                <ds-flex>
+                  <ds-flex-item>
+                    <label>{{ category.name }}</label>
+                  </ds-flex-item>
+                  <ds-space margin-bottom="xx-large" />
+                </ds-flex>
+              </ds-flex>
+            </ds-flex>
+          </ds-flex-item>
+        </ds-flex>
+      </ds-container>
+    </template>
+  </dropdown>
+</template>
+<script>
+import Dropdown from '~/components/Dropdown'
+import { filterPosts } from '~/graphql/PostQuery.js'
+import { mapMutations } from 'vuex'
+
+export default {
+  components: {
+    Dropdown,
+  },
+  props: {
+    placement: { type: String, default: 'bottom-start' },
+    offset: { type: [String, Number], default: '16' },
+    categories: { type: Array, default: () => [] },
+  },
+  data() {
+    return {
+      pageSize: 12,
+      selectedCategoryIds: [],
+      allCategories: true,
+    }
+  },
+  computed: {
+    chunk() {
+      const groupedCategories = []
+      for (let i = 0; i < this.categories.length; i++) {
+        const last = groupedCategories[groupedCategories.length - 1]
+        if (!last || last.length === 2) {
+          groupedCategories.push([this.categories[i]])
+        } else {
+          last.push(this.categories[i])
+        }
+      }
+      return groupedCategories
+    },
+  },
+  methods: {
+    ...mapMutations({
+      setPosts: 'posts/SET_POSTS',
+    }),
```


setting `Posts` in `Vuex` allows us to be able to change the `Posts` in the newsfeed without needing to use a child component.

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-07-09T23:48:12Z" title="Wednesday, July 10th 2019, 1:48:12 am +02:00">Jul 10, 2019</time>_

---


I don't quite get the reason for using Vuex here, could you explain that in more detail?

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-10T18:28:23Z" title="Wednesday, July 10th 2019, 8:28:23 pm +02:00">Jul 10, 2019</time>_

---


Well, I think I can, yes. The first thing I tried is to add the filter to the `FilterPosts.vue` and then update the apollo cache when the query returned successfully, but I had several issues.

One is that when I added `this.$apollo.query` and tried to update the stores cache it didn't update the `Posts` at all. I was able to get it to work with `this.$apollo.mutate`, even though it is actually a query and doesn't mutate anything, but it only worked if the user double-clicked on the category in the dropdown, which sent two network requests, but updated the `Posts` on the second.

I wanted to make use of the `Post` query with built-in filter that I think you added as part of the `Your filter bubble`, but since `FilterPosts` is a child of `default.vue`, and not a child of `webapp/pages/index.vue`, it's not possible to call `changeFilterBubble(filter)` without using global even listeners, which we've decided to refactor out and not use going forward.

I placed the filter posts dropdown in the `FilterMenu.vue`, and got it working fine, but in a pairing session with @alina-beck, she gave the feedback that it is probably better in the navbar, as filtering is a kind of searching, and it might not be clear what it is in the `FilterMenu`. I tend to agree with her.

Enter Vuex, I am quite familiar with using Redux as a way of managing state and passing props from related but not parent-child components, so I refactored the code to introduce the store where we can update the store after filtering for `Posts`, which in turn updates the `UI`.